library/nist_kw.c

```diff
@@ -0,0 +1,709 @@
+    *out_len = 0;
+
+    if( semiblocks < MIN_SEMIBLOCKS_COUNT )
+    {
+        ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+        goto cleanup;
+    }
+
+    memcpy( output, input, semiblocks * KW_SEMIBLOCK_LENGTH );
+
+    /* Calculate intermediate values */
+    for( t = s; t >= 1; t-- )
+    {
+        calc_a_xor_t( A, t );
+
+        memcpy( inbuff, A, KW_SEMIBLOCK_LENGTH );
+        memcpy( inbuff + KW_SEMIBLOCK_LENGTH, R, KW_SEMIBLOCK_LENGTH );
+
+        ret = mbedtls_cipher_update( &ctx->cipher_ctx,
+                                     inbuff, 16, outbuff, &olen );
+        if( ret != 0 )
+            goto cleanup;
+
+        memcpy( A, outbuff, KW_SEMIBLOCK_LENGTH );
+
+        /* Set R as LSB64 of outbuff */
+        memcpy( R, outbuff + KW_SEMIBLOCK_LENGTH, KW_SEMIBLOCK_LENGTH );
+        R -= KW_SEMIBLOCK_LENGTH;
+        if( R < output + KW_SEMIBLOCK_LENGTH )
+            R = output + ( semiblocks -1 ) * KW_SEMIBLOCK_LENGTH;
+
+    }
+
+    *out_len = semiblocks * KW_SEMIBLOCK_LENGTH;
+
+cleanup:
+    if( ret != 0)
+        memset( output, 0, semiblocks * KW_SEMIBLOCK_LENGTH );
+    mbedtls_platform_zeroize( inbuff, KW_SEMIBLOCK_LENGTH * 2 );
+    mbedtls_platform_zeroize( outbuff, KW_SEMIBLOCK_LENGTH * 2 );
+
+    return( ret );
+}
+
+/*
+ * KW-AD as defined in SP 800-38F section 6.2
+ * KWP-AD as defined in SP 800-38F section 6.3
+ */
+int mbedtls_nist_kw_unwrap( mbedtls_nist_kw_context *ctx,
+                            mbedtls_nist_kw_mode_t mode,
+                            const unsigned char *input, size_t in_len,
+                            unsigned char *output, size_t *out_len )
+{
+    int ret = 0;
+    size_t i;
+    const uint32_t ICV1[] = NIST_KW_ICV1;
+    const uint32_t ICV2[] = NIST_KW_ICV2;
+    *out_len = 0;
+
+    if( mode == MBEDTLS_KW_MODE_KW )
+    {
+        /*
+         * According to SP800-38F Table 1, the ciphertext length for KW
+         * must be between 3 to 2^54 semiblocks inclusive.
+         */
+        if( in_len < 24
+#if SIZE_MAX > 0x200000000000000
+            || in_len > 0x200000000000000
+#endif
+            )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+        ret = unwrap( ctx, input, in_len / KW_SEMIBLOCK_LENGTH,
+                      output, out_len );
+        if( ret != 0 )
+            goto cleanup;
+
+        if ( memcmp( ICV1, output, KW_SEMIBLOCK_LENGTH ) != 0 )
+        {
+            ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
+            goto cleanup;
+        }
+
+        /*
+         * shift output to point to P
+         */
+        for( i = 0; i < ( *out_len / KW_SEMIBLOCK_LENGTH ); i++ )
+        {
+            memcpy( output + ( i * KW_SEMIBLOCK_LENGTH ),
+                    output + ( ( i + 1 ) * KW_SEMIBLOCK_LENGTH ),
+                    KW_SEMIBLOCK_LENGTH );
+        }
+        *out_len = *out_len - KW_SEMIBLOCK_LENGTH;
+    }
+    else // KWP
+    {
+        int padlen = 0;
+        uint32_t Plen;
```

Comment:
Unless the standard literally uses "Plen", please use Mbed TLS naming conventions for variables.

Reply:
The standard uses `Plen`
See section 6.3 step 5 of the algorithm